driver-core/src/main/com/mongodb/internal/connection/CommandMessage.java

```diff
@@ -107,30 +111,49 @@ public final class CommandMessage extends RequestMessage {
         this.serverApi = serverApi;
     }
 
+    // Create a BsonDocument representing the logical document encoded by an OP_MSG.
+    // The returned document will contain all the fields, including ones represented by
+    // OP_MSG Sections of type Kind 1: Document Sequence
     BsonDocument getCommandDocument(final ByteBufferBsonOutput bsonOutput) {
-        ByteBufBsonDocument byteBufBsonDocument = ByteBufBsonDocument.createOne(bsonOutput,
-                getEncodingMetadata().getFirstDocumentPosition());
-        BsonDocument commandBsonDocument;
-
-        if (containsPayload()) {
-            commandBsonDocument = byteBufBsonDocument.toBaseBsonDocument();
-
-            int payloadStartPosition = getEncodingMetadata().getFirstDocumentPosition()
-                    + byteBufBsonDocument.getSizeInBytes()
-                    + 1 // payload type
-                    + 4 // payload size
-                    + payload.getPayloadName().getBytes(StandardCharsets.UTF_8).length + 1;  // null-terminated UTF-8 payload name
-            commandBsonDocument.append(payload.getPayloadName(),
-                    new BsonArray(ByteBufBsonDocument.createList(bsonOutput, payloadStartPosition)));
-        } else {
-            commandBsonDocument = byteBufBsonDocument;
+        List<ByteBuf> duplicateByteBuffers = bsonOutput.getByteBuffers();
+        try {
+            CompositeByteBuf outputByteBuf = new CompositeByteBuf(duplicateByteBuffers);
+            outputByteBuf.position(getEncodingMetadata().getFirstDocumentPosition());
+            ByteBufBsonDocument byteBufBsonDocument = createOne(outputByteBuf);
+
+            if (outputByteBuf.hasRemaining()) {
+                BsonDocument commandBsonDocument = byteBufBsonDocument.toBaseBsonDocument();
+
+                while (outputByteBuf.hasRemaining()) {
+                    outputByteBuf.position(outputByteBuf.position() + 1 /* payload type */ + 4 /* payload size */);
+                    String payloadName = getPayloadName(outputByteBuf);
```

Comment:
Now we get the payload name directly from the OP_MSG instead of from the SplittablePayload, making it easier to support any number of document sequence independent of any future changes to SplittablePayload

Reply:
If we now get the "payload name" (which is actually, the [sequence identifier](https://www.mongodb.com/docs/manual/reference/mongodb-wire-protocol/#kind-1--document-sequence)), why do we still use `SplittablePayload.getPayloadName`, including using it in `CommandMessage`?

I don't yet have a good understanding of what's going on, but the above seems quite confusing.

Reply:
When writing the OP_MSG Kind 1 Section, we need to get the payload name from somewhere.  It comes from `SplittablePayload`.  This code is just getting it out of the OP_MSG after it's been written, so it no longer has to depend on `SplittablePayload`, which is now free to evolve independent of this method (for example, to support multiple payloads).

Reply:
If you are suggesting that the `getPayloadName` method should be renamed to something like `getSequenceIdentifier`, I agree.

Reply:
This all goes a long way back, to when there were many more subclasses of RequestMessage (we used to have subclasses for OP_INSERT, OP_UPDATE, OP_DELETE, etc).  But now, there is only CommandMessage, and except for the annoyance of having to use OP_QUERY for the very first message we send on a connection, CommandMessage only encodes OP_MSG.  Which makes `EncodingMetadata#getFirstDocumentPosition` effectively a constant.  

When we removed support for all the other op codes, we simplified a lot, but there is more that can be done.  Just not sure it should be done in this PR.

Reply:
Thank you, I now understand why it was varying previously, but not the reasons behind the design that I described previously. I created https://jira.mongodb.org/browse/JAVA-5554 about improving this.